pyanaconda/payload/dnfpayload.py

```diff
@@ -412,80 +414,149 @@ def addRepo(self, ksrepo):
+            # only use type specififcations if --nodefault,
+            # --optional or both are used
+            if not default or optional:
+                types = "mandatory"
+                if default:
+                    types = "{},default".format(types)
+                if optional:
+                    types = "{},optional".format(types)
+
+                group_spec = "@{group_name}/{types}".format(group_name=group.name, types=types)
+            else:
+                # if group is a regular group this is equal to @group/mandatory,default
+                group_spec = "@{group_name}".format(group_name=group.name)
 
+            group_specs.add(group_spec)
+
+        # handle packages
+        selected_packages = set()
+        excluded_packages = set()
         for pkg_name in self.data.packages.excludedList:
-            self._exclude_package(pkg_name)
             log.info("excluded package: '%s'", pkg_name)
+            excluded_packages.add(pkg_name)
 
         for pkg_name in self.data.packages.packageList:
-            try:
-                self._install_package(pkg_name)
-                log.info("selected package: '%s'", pkg_name)
-            except payload.NoSuchPackage as e:
-                self._miss(e)
+            log.info("selected package: '%s'", pkg_name)
+            selected_packages.add(pkg_name)
+
+        kernel_package = self._get_kernel_package()
+        if kernel_package:
+            selected_packages.add(kernel_package)
+
+        # resolve packages and groups required by Anaconda
+        self.requirements.apply()
 
-        self._select_kernel_package()
+        # let's piece it all together and feed it to DNF at once
+        include_set = set()
+        exclude_set = set()
+        # note about formatting:
+        # - leading @ signifies a group or module
+        # - no leading @ means a package
+
+        # add environment
+        if env:
+            include_set.add("@^{}".format(env))
+
+        # add groups
+        include_set.update(group_specs)
+        for group_name in self._req_groups:
+            include_set.add("@{}".format(group_name))
+        # add packages
+        include_set.update(selected_packages)
+        include_set.update(self._req_packages)
+
+        # exclude groups
+        exclude_set.update(excluded_group_specs)
+        # exclude packages
+        exclude_set.update(excluded_packages)
+
+        # log the resulting set
+        log.debug("transaction include set")
+        log.debug(include_set)
+        log.debug("transaction exclude set")
+        log.debug(exclude_set)
+
+        # feed it to DNF
+        try:
+            self._base.install_specs(install=include_set, exclude=exclude_set)
+        except Exception as e:  # pylint: disable=broad-except
+            exception_message = str(e)
+            self._payload_setup_error(e)
 
     def _apply_requirements(self, requirements):
+        self._req_groups = set()
+        self._req_packages = set()
```

Comment:
The solution with a callback is not useful anymore. Ideally, this callback could be replaced with two methods `get_required_packages` and `get_required_groups` that could return a set of packages and a set of groups, so we can avoid these attributes.

Reply:
+1, we should just think if/how we want to keep the possibility of checking if (all the) the requirements were actually applied by the payload

Reply:
Definitely not looking nice at the moment. But on the other hand there are already many changes in this PR. So I'd personally prefer to handle this better in a followup PR.

Reply:
Ok.